toolchain/check/eval.cpp

```diff
@@ -1895,6 +1898,58 @@ static auto IsPeriodSelf(EvalContext& eval_context, SemIR::ConstantId const_id)
   return false;
 }
 
+template <>
+auto TryEvalTypedInst<SemIR::LookupImplWitness>(EvalContext& eval_context,
+                                                SemIR::InstId inst_id,
+                                                SemIR::Inst inst)
+    -> SemIR::ConstantId {
+  CARBON_CHECK(inst_id.has_value());
+
+  Phase phase = Phase::Concrete;
+
+  CARBON_CHECK(
+      ReplaceTypeWithConstantValue(eval_context, inst_id, &inst, &phase));
+  CARBON_CHECK(ReplaceAllFieldsWithConstantValues(eval_context, &inst, &phase));
+
+  if (!eval_context.has_specific()) {
+    return ConvertEvalResultToConstantId(
+        eval_context.context(), ConstantEvalResult::NewSamePhase(inst), phase);
+  } else {
+    auto lookup_inst = inst.As<SemIR::LookupImplWitness>();
+
+    // The self value is canonicalized in order to produce a canonical
+    // LookupImplWitness instruction. We save the non-canonical instruction as
+    // it may be a concrete `FacetValue` that contains a concrete witness.
+    auto non_canonical_query_self_inst_id = lookup_inst.query_self_inst_id;
+    lookup_inst.query_self_inst_id = GetCanonicalizedFacetOrTypeValue(
+        eval_context.context(), lookup_inst.query_self_inst_id);
+
+    auto result = EvalLookupSingleImplWitness(
+        eval_context.context(), SemIR::LocId(inst_id), lookup_inst,
+        non_canonical_query_self_inst_id,
+        /*poison_concrete_results=*/true);
+    // The `LookupImplWitness` instruction is only created and evaluated when we
+    // already know there is a witness.
+    CARBON_CHECK(result.has_value());
+    if (!result.has_concrete_value()) {
+      return ConvertEvalResultToConstantId(
+          eval_context.context(), ConstantEvalResult::NewSamePhase(lookup_inst),
+          phase);
+    }
+    return ConvertEvalResultToConstantId(
+        eval_context.context(),
+        ConstantEvalResult::Existing(
+            eval_context.constant_values().Get(result.concrete_witness())),
+        phase);
+  }
+
+  return ConvertEvalResultToConstantId(
+      eval_context.context(),
+      EvalConstantInst(eval_context.context(), inst_id,
+                       inst.As<SemIR::LookupImplWitness>()),
```

Comment:
I'm a bit confused here:  it looks like you deleted the function that you're calling here. Does this link?

Reply:
It linked, but yes this was indeed - it was dead code though so I guess I got away with it. The next commits restored that function and made this call into not dead code.

Reply:
FWIW one thing I considered was adding another flavour of EvalConstantInst which takes additional information (the current context specific id, would work) to avoid this eval specialization.